File: telegram/utils/currency_utils.py
"""Utilitaires pour les devises"""
import os
from config.currencies import SUPPORTED_CURRENCIES

def get_country_from_currency(currency):
    """Obtenir pays depuis devise"""
    currency_info = SUPPORTED_CURRENCIES.get(currency)
    return currency_info['country'] if currency_info else 'US'

def requires_decimals(currency):
    """Vérifier si devise nécessite décimales"""
    no_decimal_currencies = ['JPY', 'KRW']
    return currency not in no_decimal_currencies

def to_stripe_amount(amount, currency):
    """Convertir pour Stripe (centimes)"""
    if requires_decimals(currency):
        return int(amount * 100)
    return int(amount)

def from_stripe_amount(stripe_amount, currency):
    """Convertir depuis Stripe"""
    if requires_decimals(currency):
        return stripe_amount / 100
    return stripe_amount

def get_currency_symbol(currency):
    """Obtenir symbole devise"""
    currency_info = SUPPORTED_CURRENCIES.get(currency)
    return currency_info['symbol'] if currency_info else currency

def is_valid_currency(currency):
    """Valider code devise"""
    return currency in SUPPORTED_CURRENCIES

def calculate_platform_fee(amount, fee_percentage=None):
    """Calculer frais plateforme"""
    if fee_percentage is None:
        fee_percentage = float(os.getenv('PLATFORM_FEE_PERCENTAGE', 10))
    return amount * (fee_percentage / 100)

File: backend/utils/currencyUtils.js
// Utilitaires pour les devises

// Obtenir le pays à partir du code devise
const getCountryFromCurrency = (currency) => {
  const currencyToCountry = {
    'USD': 'US',
    'EUR': 'FR', // Ou tout pays de la zone Euro
    'JPY': 'JP',
    'GBP': 'GB',
    'AUD': 'AU',
    'CAD': 'CA',
    'CHF': 'CH',
    'CNY': 'CN',
    'SGD': 'SG',
    'SEK': 'SE',
    'NOK': 'NO',
    'KRW': 'KR',
    'BRL': 'BR'
  };
  
  return currencyToCountry[currency] || 'US';
};

// Calculer les frais de plateforme
const calculatePlatformFee = (amount, feePercentage = 10) => {
  return Math.round(amount * (feePercentage / 100));
};

// Vérifier si une devise nécessite des décimales
const requiresDecimals = (currency) => {
  const noDecimalCurrencies = ['JPY', 'KRW'];
  return !noDecimalCurrencies.includes(currency);
};

// Convertir un montant pour Stripe (en centimes)
const toStripeAmount = (amount, currency) => {
  if (requiresDecimals(currency)) {
    return Math.round(amount * 100);
  }
  return Math.round(amount);
};

// Convertir depuis Stripe (depuis centimes)
const fromStripeAmount = (stripeAmount, currency) => {
  if (requiresDecimals(currency)) {
    return stripeAmount / 100;
  }
  return stripeAmount;
};

// Obtenir le symbole de devise
const getCurrencySymbol = (currency) => {
  const symbols = {
    USD: '$',
    EUR: '€',
    JPY: '¥',
    GBP: '£',
    AUD: 'A$',
    CAD: 'C$',
    CHF: 'CHF',
    CNY: '¥',
    SGD: 'S$',
    SEK: 'kr',
    NOK: 'kr',
    KRW: '₩',
    BRL: 'R$'
  };
  
  return symbols[currency] || currency;
};

// Valider un code devise
const isValidCurrency = (currency) => {
  const validCurrencies = ['USD', 'EUR', 'JPY', 'GBP', 'AUD', 'CAD', 
                          'CHF', 'CNY', 'SGD', 'SEK', 'NOK', 'KRW', 'BRL'];
  return validCurrencies.includes(currency);
};

module.exports = {
  getCountryFromCurrency,
  calculatePlatformFee,
  requiresDecimals,
  toStripeAmount,
  fromStripeAmount,
  getCurrencySymbol,
  isValidCurrency
};